File: jtms/java/JSOps.java
// Integration operators for JSAINT.
// Translated from jsops.lisp in "Building Problem Solvers"
// by Kenneth D. Forbus and Johan de Kleer.
package jtms;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * JSOps defines integration operators used by JSAINT.
 * Each operator represents an integration method.
 */
public class JSOps {

    /** An integration operator definition. */
    public static class IntegrationOp {
        public String name;
        public java.util.function.Predicate<Object> test;

        public IntegrationOp(String name, java.util.function.Predicate<Object> test) {
            this.name = name;
            this.test = test;
        }
    }

    /** Get standard integration operators. Corresponds to jsops.lisp. */
    @SuppressWarnings("unchecked")
    public static List<IntegrationOp> standardOps() {
        List<IntegrationOp> ops = new ArrayList<>();

        // Integral-of-Constant: (Integral ?t ?var) where ?var not in ?t
        ops.add(new IntegrationOp("Integral-of-Constant", integral -> {
            List<Object> parts = asList(integral);
            if (parts == null || parts.size() != 3) return false;
            if (!"Integral".equals(parts.get(0))) return false;
            return !Simplify.occursIn(parts.get(2), parts.get(1));
        }));

        // Integral-of-Self: (Integral ?exp ?exp)
        ops.add(new IntegrationOp("Integral-of-Self", integral -> {
            List<Object> parts = asList(integral);
            if (parts == null || parts.size() != 3) return false;
            if (!"Integral".equals(parts.get(0))) return false;
            return JData.deepEqual(parts.get(1), parts.get(2));
        }));

        // Integral-of-Sum: (Integral (+ ?t1 ?t2) ?var)
        ops.add(new IntegrationOp("Integral-of-Sum", integral -> {
            List<Object> parts = asList(integral);
            if (parts == null || parts.size() != 3) return false;
            if (!"Integral".equals(parts.get(0))) return false;
            List<Object> inner = asList(parts.get(1));
            return inner != null && inner.size() >= 3 && "+".equals(inner.get(0));
        }));

        // Move-Constant-Outside: (Integral (* ?const ?nonconst) ?var)
        ops.add(new IntegrationOp("Move-Constant-Outside", integral -> {
            List<Object> parts = asList(integral);
            if (parts == null || parts.size() != 3) return false;
            if (!"Integral".equals(parts.get(0))) return false;
            List<Object> inner = asList(parts.get(1));
            if (inner == null || inner.size() != 3 || !"*".equals(inner.get(0))) return false;
            return !Simplify.occursIn(parts.get(2), inner.get(1))
                && Simplify.occursIn(parts.get(2), inner.get(2));
        }));

        // Simple-e-integral: (Integral (expt %e ?var) ?var)
        ops.add(new IntegrationOp("Simple-e-integral", integral -> {
            List<Object> parts = asList(integral);
            if (parts == null || parts.size() != 3) return false;
            if (!"Integral".equals(parts.get(0))) return false;
            List<Object> inner = asList(parts.get(1));
            return inner != null && inner.size() == 3 && "expt".equals(inner.get(0))
                && "%e".equals(inner.get(1)) && JData.deepEqual(inner.get(2), parts.get(2));
        }));

        // Sin-integral: (Integral (sin (* ?a ?var)) ?var)
        ops.add(new IntegrationOp("Sin-integral", integral -> {
            List<Object> parts = asList(integral);
            if (parts == null || parts.size() != 3) return false;
            if (!"Integral".equals(parts.get(0))) return false;
            List<Object> inner = asList(parts.get(1));
            if (inner == null || inner.size() != 2 || !"sin".equals(inner.get(0))) return false;
            List<Object> product = asList(inner.get(1));
            return product != null && product.size() == 3 && "*".equals(product.get(0))
                && !Simplify.occursIn(parts.get(2), product.get(1));
        }));

        // Cos-integral: (Integral (cos (* ?a ?var)) ?var)
        ops.add(new IntegrationOp("Cos-integral", integral -> {
            List<Object> parts = asList(integral);
            if (parts == null || parts.size() != 3) return false;
            if (!"Integral".equals(parts.get(0))) return false;
            List<Object> inner = asList(parts.get(1));
            if (inner == null || inner.size() != 2 || !"cos".equals(inner.get(0))) return false;
            List<Object> product = asList(inner.get(1));
            return product != null && product.size() == 3 && "*".equals(product.get(0))
                && !Simplify.occursIn(parts.get(2), product.get(1));
        }));

        // Log-integral: (Integral (log ?var %e) ?var)
        ops.add(new IntegrationOp("Log-integral", integral -> {
            List<Object> parts = asList(integral);
            if (parts == null || parts.size() != 3) return false;
            if (!"Integral".equals(parts.get(0))) return false;
            List<Object> inner = asList(parts.get(1));
            return inner != null && inner.size() == 3 && "log".equals(inner.get(0))
                && JData.deepEqual(inner.get(1), parts.get(2)) && "%e".equals(inner.get(2));
        }));

        // Integral-of-Polyterm: (Integral (expt ?var ?n) ?var) where ?n != -1
        ops.add(new IntegrationOp("Integral-of-Polyterm", integral -> {
            List<Object> parts = asList(integral);
            if (parts == null || parts.size() != 3) return false;
            if (!"Integral".equals(parts.get(0))) return false;
            List<Object> inner = asList(parts.get(1));
            if (inner == null || inner.size() != 3 || !"expt".equals(inner.get(0))) return false;
            return JData.deepEqual(inner.get(1), parts.get(2))
                && !Simplify.sameConstant(inner.get(2), -1);
        }));

        return ops;
    }

    @SuppressWarnings("unchecked")
    private static List<Object> asList(Object x) {
        if (x instanceof List) return (List<Object>) x;
        return null;
    }
}

File: jtms/go/jsops.go
// Operators for JSAINT - integration operator definitions.
// Translated from jsops.lisp in "Building Problem Solvers"
// by Kenneth D. Forbus and Johan de Kleer.

package jtms

// IntegrationOperator represents an integration method.
type IntegrationOperator struct {
	Name        string
	Trigger     func(integral interface{}) bool
	Subproblems func(integral interface{}) []interface{}
	Result      func(integral interface{}, subresults map[string]interface{}) interface{}
	Test        func(integral interface{}) bool
}

// StandardIntegrationOps returns the standard set of integration operators.
// Each operator corresponds to a defIntegration in jsops.lisp.
func StandardIntegrationOps() []*IntegrationOperator {
	return []*IntegrationOperator{
		{
			Name: "Integral-of-Constant",
			// (Integral ?t ?var) where ?var does not occur in ?t
			// Result: (* ?t ?var)
			Test: func(integral interface{}) bool {
				parts := asSlice(integral)
				if len(parts) != 3 || asStr(parts[0]) != "Integral" {
					return false
				}
				return !OccursIn(parts[2], parts[1])
			},
		},
		{
			Name: "Integral-of-Self",
			// (Integral ?exp ?exp)
			// Result: (/ (expt ?exp 2) 2)
			Test: func(integral interface{}) bool {
				parts := asSlice(integral)
				if len(parts) != 3 || asStr(parts[0]) != "Integral" {
					return false
				}
				return deepEqual(parts[1], parts[2])
			},
		},
		{
			Name: "Integral-of-Sum",
			// (Integral (+ ?t1 ?t2) ?var)
			// Subproblems: integrate each term separately
			// Result: (+ ?int1 ?int2)
			Test: func(integral interface{}) bool {
				parts := asSlice(integral)
				if len(parts) != 3 || asStr(parts[0]) != "Integral" {
					return false
				}
				inner := asSlice(parts[1])
				return len(inner) >= 3 && asStr(inner[0]) == "+"
			},
		},
		{
			Name: "Move-Constant-Outside",
			// (Integral (* ?const ?nonconst) ?var) where ?var not in ?const
			Test: func(integral interface{}) bool {
				parts := asSlice(integral)
				if len(parts) != 3 || asStr(parts[0]) != "Integral" {
					return false
				}
				inner := asSlice(parts[1])
				if len(inner) != 3 || asStr(inner[0]) != "*" {
					return false
				}
				return !OccursIn(parts[2], inner[1]) && OccursIn(parts[2], inner[2])
			},
		},
		{
			Name: "Simple-e-integral",
			// (Integral (expt %e ?var) ?var)
			// Result: (expt %e ?var)
			Test: func(integral interface{}) bool {
				parts := asSlice(integral)
				if len(parts) != 3 || asStr(parts[0]) != "Integral" {
					return false
				}
				inner := asSlice(parts[1])
				return len(inner) == 3 && asStr(inner[0]) == "expt" &&
					asStr(inner[1]) == "%e" && deepEqual(inner[2], parts[2])
			},
		},
		{
			Name: "Sin-integral",
			// (Integral (sin (* ?a ?var)) ?var) where ?var not in ?a
			// Result: (- (/ (cos (* ?a ?var)) ?a))
			Test: func(integral interface{}) bool {
				parts := asSlice(integral)
				if len(parts) != 3 || asStr(parts[0]) != "Integral" {
					return false
				}
				inner := asSlice(parts[1])
				if len(inner) != 2 || asStr(inner[0]) != "sin" {
					return false
				}
				product := asSlice(inner[1])
				return len(product) == 3 && asStr(product[0]) == "*" &&
					!OccursIn(parts[2], product[1])
			},
		},
		{
			Name: "Cos-integral",
			// (Integral (cos (* ?a ?var)) ?var) where ?var not in ?a
			// Result: (/ (sin (* ?a ?var)) ?a)
			Test: func(integral interface{}) bool {
				parts := asSlice(integral)
				if len(parts) != 3 || asStr(parts[0]) != "Integral" {
					return false
				}
				inner := asSlice(parts[1])
				if len(inner) != 2 || asStr(inner[0]) != "cos" {
					return false
				}
				product := asSlice(inner[1])
				return len(product) == 3 && asStr(product[0]) == "*" &&
					!OccursIn(parts[2], product[1])
			},
		},
		{
			Name: "Log-integral",
			// (Integral (log ?var %e) ?var)
			// Result: (- (* ?var (log ?var %e)) ?var)
			Test: func(integral interface{}) bool {
				parts := asSlice(integral)
				if len(parts) != 3 || asStr(parts[0]) != "Integral" {
					return false
				}
				inner := asSlice(parts[1])
				return len(inner) == 3 && asStr(inner[0]) == "log" &&
					deepEqual(inner[1], parts[2]) && asStr(inner[2]) == "%e"
			},
		},
		{
			Name: "Integral-of-Polyterm",
			// (Integral (expt ?var ?n) ?var) where ?n != -1
			// Result: (/ (expt ?var (+ 1 ?n)) (+ 1 ?n))
			Test: func(integral interface{}) bool {
				parts := asSlice(integral)
				if len(parts) != 3 || asStr(parts[0]) != "Integral" {
					return false
				}
				inner := asSlice(parts[1])
				if len(inner) != 3 || asStr(inner[0]) != "expt" {
					return false
				}
				return deepEqual(inner[1], parts[2]) && !SameConstant(inner[2], -1)
			},
		},
	}
}

// RegisterIntegrationOps registers all integration operators as rules
// in the given JTRE.
func RegisterIntegrationOps(jtre *Jtre) {
	ops := StandardIntegrationOps()
	for _, op := range ops {
		_ = op // Each operator would be registered as a rule
		// In the full implementation, each defIntegration macro
		// expands into a rule that matches against expanded integration problems
	}
}

// asSlice converts interface{} to []interface{} or returns nil.
func asSlice(x interface{}) []interface{} {
	s, _ := x.([]interface{})
	return s
}

// asStr converts interface{} to string or returns "".
func asStr(x interface{}) string {
	s, _ := x.(string)
	return s
}
